Zend/zend_lazy_objects.c

```diff
@@ -0,0 +1,705 @@
+			ZEND_ASSERT(EG(exception));
+			return NULL;
+		}
+	}
+
+	zval *default_properties_table = CE_DEFAULT_PROPERTIES_TABLE(ce);
+	zval *properties_table = obj->properties_table;
+
+	OBJ_EXTRA_FLAGS(obj) &= ~(IS_OBJ_LAZY|IS_OBJ_LAZY_PROXY);
+
+	for (int i = 0; i < ce->default_properties_count; i++) {
+		if (Z_PROP_FLAG_P(&properties_table[i]) & IS_PROP_LAZY) {
+			ZVAL_COPY_PROP(&properties_table[i], &default_properties_table[i]);
+		}
+	}
+
+	zend_lazy_object_del_info(obj);
+
+	return obj;
+}
+
+/* Revert initializer effects */
+static void zend_lazy_object_revert_init(zend_object *obj, zval *properties_table_snapshot, HashTable *properties_snapshot)
+{
+	zend_class_entry *ce = obj->ce;
+
+	if (ce->default_properties_count) {
+		ZEND_ASSERT(properties_table_snapshot);
+		zval *properties_table = obj->properties_table;
+
+		for (int i = 0; i < ce->default_properties_count; i++) {
+			zval *p = &properties_table[i];
+			zend_object_dtor_property(obj, p);
+			ZVAL_COPY_PROP(p, &properties_table_snapshot[i]);
+			Z_TRY_DELREF_P(p);
+
+			zend_property_info *prop_info = ce->properties_info_table[i];
+			if (Z_ISREF_P(p) && prop_info && ZEND_TYPE_IS_SET(prop_info->type)) {
+				ZEND_REF_ADD_TYPE_SOURCE(Z_REF_P(p), prop_info);
+			}
+		}
+
+		efree(properties_table_snapshot);
+	}
+	if (properties_snapshot) {
+		if (obj->properties != properties_snapshot) {
+			ZEND_ASSERT((GC_FLAGS(properties_snapshot) & IS_ARRAY_IMMUTABLE) || GC_REFCOUNT(properties_snapshot) >= 1);
+			zend_release_properties(obj->properties);
+			obj->properties = properties_snapshot;
+		} else {
+			ZEND_ASSERT((GC_FLAGS(properties_snapshot) & IS_ARRAY_IMMUTABLE) || GC_REFCOUNT(properties_snapshot) > 1);
+			zend_release_properties(properties_snapshot);
+		}
+	} else if (obj->properties) {
+		zend_release_properties(obj->properties);
+		obj->properties = NULL;
+	}
+
+	OBJ_EXTRA_FLAGS(obj) |= IS_OBJ_LAZY;
+}
+
+static bool zend_lazy_object_compatible(zend_object *real_object, zend_object *lazy_object)
+{
+	if (EXPECTED(real_object->ce == lazy_object->ce)) {
+		return true;
+	}
+
+	if (!instanceof_function(lazy_object->ce, real_object->ce)) {
+		return false;
+	}
+
+	/* zend_hash_num_elements(ce.properties_info) reports the actual number of
+	 * properties. ce.default_properties_count is off by the number of property
+	 * overrides. */
+	if (zend_hash_num_elements(&lazy_object->ce->properties_info) != zend_hash_num_elements(&real_object->ce->properties_info)) {
+		return false;
+	}
+
+	return lazy_object->ce->destructor == real_object->ce->destructor
+		&& lazy_object->ce->clone == real_object->ce->clone;
+}
+
+/* Initialize a lazy proxy object */
+static zend_object *zend_lazy_object_init_proxy(zend_object *obj)
+{
+	ZEND_ASSERT(zend_object_is_lazy_proxy(obj));
+	ZEND_ASSERT(!zend_lazy_object_initialized(obj));
+
+	zend_lazy_object_info *info = zend_lazy_object_get_info(obj);
+
+	/* prevent reentrant initialization */
+	OBJ_EXTRA_FLAGS(obj) &= ~(IS_OBJ_LAZY|IS_OBJ_LAZY_PROXY);
```

Comment:
This must remove the `IS_OBJ_LAZY_PROXY` flag, otherwise it may try to fetch the real instance when a field is accessed during initialization, but it doesn't exist yet